File: acutis.api/Acutis.Api/Services/Residents/ResidentService.cs
using Acutis.Api.Contracts;
using Acutis.Domain.Entities;
using Acutis.Infrastructure.Data;
using Microsoft.EntityFrameworkCore;

namespace Acutis.Api.Services.Residents;

public interface IResidentService
{
    Task<IReadOnlyList<ResidentListItemDto>> GetAllResidentsAsync(
        string unitId,
        CancellationToken cancellationToken = default);

    Task<RecordDischargeResponse> RecordDischargeAsync(
        Guid residentGuid,
        RecordDischargeRequest request,
        Guid actorUserId,
        CancellationToken cancellationToken = default);
}

public sealed class ResidentService : IResidentService
{
    private readonly AcutisDbContext _dbContext;

    public ResidentService(AcutisDbContext dbContext)
    {
        _dbContext = dbContext;
    }

    public async Task<IReadOnlyList<ResidentListItemDto>> GetAllResidentsAsync(
        string unitId,
        CancellationToken cancellationToken = default)
    {
        var normalizedUnit = NormalizeUnit(unitId);
        var unit = await _dbContext.Units
            .AsNoTracking()
            .Where(x => x.Code == normalizedUnit)
            .Select(x => new { x.Id, x.Code })
            .FirstOrDefaultAsync(cancellationToken);

        var activeEpisodes = unit is null
            ? []
            : await _dbContext.ResidentProgrammeEpisodes
                .AsNoTracking()
                .Where(x => x.UnitId == unit.Id && x.EndDate == null)
                .OrderByDescending(x => x.StartDate)
                .Select(x => new
                {
                    x.Id,
                    x.ResidentId,
                    x.ResidentCaseId,
                    x.UnitId,
                    x.StartDate,
                    x.CurrentWeekNumber,
                    x.CentreEpisodeCode,
                    x.EntryYear,
                    x.EntryWeek,
                    x.EntrySequence,
                    x.RoomNumber,
                    x.ExpectedCompletionDate,
                    x.PrimaryAddiction
                })
                .ToListAsync(cancellationToken);

        var currentEpisodeByResident = activeEpisodes
            .GroupBy(x => x.ResidentId)
            .ToDictionary(x => x.Key, x => x.First());

        var residentIdsWithEpisode = currentEpisodeByResident.Keys.ToHashSet();

        var residentsFromEpisode = residentIdsWithEpisode.Count == 0
            ? []
            : await _dbContext.Residents
                .AsNoTracking()
                .Where(resident => residentIdsWithEpisode.Contains(resident.Id))
                .ToListAsync(cancellationToken);

        var residentsFromLegacyUnit = await _dbContext.Residents
            .AsNoTracking()
            .Where(resident => resident.UnitCode == normalizedUnit && !residentIdsWithEpisode.Contains(resident.Id))
            .ToListAsync(cancellationToken);

        var residents = residentsFromEpisode
            .Concat(residentsFromLegacyUnit)
            .OrderBy(resident => resident.Surname)
            .ThenBy(resident => resident.FirstName)
            .ToList();

        var caseIds = currentEpisodeByResident.Values
            .Where(x => x.ResidentCaseId.HasValue)
            .Select(x => x.ResidentCaseId!.Value)
            .Distinct()
            .ToList();

        var caseStatusById = caseIds.Count == 0
            ? new Dictionary<Guid, string>()
            : await _dbContext.ResidentCases
                .AsNoTracking()
                .Where(x => caseIds.Contains(x.Id))
                .ToDictionaryAsync(x => x.Id, x => x.CaseStatus, cancellationToken);

        return residents
            .Select(resident =>
            {
                currentEpisodeByResident.TryGetValue(resident.Id, out var episode);
                var episodeCaseStatus = episode?.ResidentCaseId is Guid residentCaseId &&
                                        caseStatusById.TryGetValue(residentCaseId, out var status)
                    ? status
                    : null;

                var currentWeekNumber = episode?.CurrentWeekNumber ?? resident.WeekNumber;
                var currentUnitGuid = episode?.UnitId ?? resident.UnitId;
                var currentUnitCode = unit?.Code ?? resident.UnitCode?.Trim().ToLowerInvariant() ?? normalizedUnit;
                var admissionDate = episode?.StartDate.ToString("yyyy-MM-dd") ?? resident.AdmissionDate?.ToString("yyyy-MM-dd");
                var roomNumber = episode?.RoomNumber?.Trim() ?? resident.RoomNumber?.Trim() ?? string.Empty;
                var expectedCompletion = episode?.ExpectedCompletionDate?.ToString("yyyy-MM-dd") ?? resident.ExpectedCompletionDate?.ToString("yyyy-MM-dd");
                var primaryAddiction = episode?.PrimaryAddiction?.Trim() ?? resident.PrimaryAddiction?.Trim() ?? string.Empty;

                return new ResidentListItemDto
                {
                Id = ToLegacyResidentId(resident.Id),
                ResidentGuid = resident.Id,
                EpisodeId = episode?.Id,
                ResidentCaseId = episode?.ResidentCaseId,
                CentreEpisodeCode = episode?.CentreEpisodeCode,
                EntryYear = episode?.EntryYear,
                EntryWeek = episode?.EntryWeek,
                EntrySequence = episode?.EntrySequence,
                CaseStatus = episodeCaseStatus,
                Psn = resident.Psn?.Trim() ?? string.Empty,
                UnitGuid = currentUnitGuid,
                FirstName = resident.FirstName?.Trim() ?? string.Empty,
                Surname = resident.Surname?.Trim() ?? string.Empty,
                Nationality = resident.Nationality?.Trim() ?? string.Empty,
                Age = resident.DateOfBirth.HasValue
                    ? CalculateAge(resident.DateOfBirth.Value.ToString("yyyy-MM-dd"))
                    : 18,
                WeekNumber = currentWeekNumber,
                RoomNumber = roomNumber,
                UnitId = currentUnitCode,
                PhotoUrl = resident.PhotoUrl,
                AdmissionDate = admissionDate,
                ExpectedCompletion = expectedCompletion,
                PrimaryAddiction = primaryAddiction,
                IsDrug = resident.IsDrug,
                IsGambeler = resident.IsGambeler,
                IsPreviousResident = resident.IsPreviousResident,
                DietaryNeedsCode = resident.DietaryNeedsCode,
                IsSnorer = resident.IsSnorer,
                HasCriminalHistory = resident.HasCriminalHistory,
                IsOnProbation = resident.IsOnProbation,
                ArgumentativeScale = resident.ArgumentativeScale,
                LearningDifficultyScale = resident.LearningDifficultyScale,
                LiteracyScale = resident.LiteracyScale
                };
            })
            .ToList();
    }

    public async Task<RecordDischargeResponse> RecordDischargeAsync(
        Guid residentGuid,
        RecordDischargeRequest request,
        Guid actorUserId,
        CancellationToken cancellationToken = default)
    {
        // Idempotency: if this clientEventId was already recorded, return it without creating a duplicate.
        var existing = await _dbContext.EpisodeEvents
            .AsNoTracking()
            .Where(e => e.ClientEventId == request.ClientEventId)
            .Select(e => new { e.Id, e.EpisodeId })
            .FirstOrDefaultAsync(cancellationToken);

        if (existing is not null)
        {
            return new RecordDischargeResponse
            {
                EpisodeEventId = existing.Id,
                EpisodeId = existing.EpisodeId,
                WasAlreadyRecorded = true
            };
        }

        var episode = await _dbContext.ResidentProgrammeEpisodes
            .Where(e => e.ResidentId == residentGuid && e.EndDate == null)
            .OrderByDescending(e => e.StartDate)
            .FirstOrDefaultAsync(cancellationToken)
            ?? throw new KeyNotFoundException($"No active episode found for resident '{residentGuid}'.");

        var exitType = (EpisodeEventType)request.ExitType;

        var episodeEvent = new EpisodeEvent
        {
            Id = Guid.NewGuid(),
            ClientEventId = request.ClientEventId,
            EpisodeId = episode.Id,
            EventType = exitType,
            EventDate = request.EventDate,
            Reason = request.Reason,
            CreatedAt = DateTime.UtcNow,
            CreatedByUserId = actorUserId,
            PayloadJson = "{}"
        };

        _dbContext.EpisodeEvents.Add(episodeEvent);

        // ExtendedStay keeps the episode open — resident is staying on past their programme end date.
        // All other exit types close the episode.
        if (exitType != EpisodeEventType.ExtendedStay)
        {
            episode.EndDate = request.EventDate;
        }

        await _dbContext.SaveChangesAsync(cancellationToken);

        return new RecordDischargeResponse
        {
            EpisodeEventId = episodeEvent.Id,
            EpisodeId = episode.Id,
            WasAlreadyRecorded = false
        };
    }

    private static int ToLegacyResidentId(Guid residentId)
    {
        var value = BitConverter.ToInt32(residentId.ToByteArray(), 0) & int.MaxValue;
        return value == 0 ? 1 : value;
    }

    private static string NormalizeUnit(string unitId)
    {
        if (string.IsNullOrWhiteSpace(unitId))
        {
            return "alcohol";
        }

        return unitId.Trim().ToLowerInvariant();
    }

    private static int CalculateAge(string dobIso)
    {
        if (!DateOnly.TryParse(dobIso, out var dob))
        {
            return 18;
        }

        var today = DateOnly.FromDateTime(DateTime.UtcNow);
        var age = today.Year - dob.Year;
        if (dob > today.AddYears(-age))
        {
            age--;
        }

        return Math.Clamp(age, 16, 120);
    }
}

File: acutis.web/src/services/residentService.ts
import { mockResidents, type Resident } from "./mockDataService";
import { UNIT_GUIDS } from "./unitIdentity";

export type AttendanceRecord = {
  residentId: number;
  present: boolean;
  reason?: string;
  description?: string;
  timestamp: string;
};

export type DischargeExitType =
  | "Completed"        // 7 - programme completion
  | "SelfDischarge"    // 8 - resident walked out
  | "ExtendedStay"     // 9 - staying on past programme end date
  | "ClinicalDischarge" // 10 - clinical or administrative decision
  | "Ejected";         // 6 - ejected from programme

const EXIT_TYPE_CODES: Record<DischargeExitType, number> = {
  Ejected: 6,
  Completed: 7,
  SelfDischarge: 8,
  ExtendedStay: 9,
  ClinicalDischarge: 10,
};

export type RecordDischargeResult = {
  episodeEventId: string;
  episodeId: string;
  wasAlreadyRecorded: boolean;
};

const store: {
  attendance: AttendanceRecord[];
} = {
  attendance: [],
};

const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

const API_BASE_URL = process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://localhost:5009";

let residentSource: "api" | "mock" = "mock";

type ResidentListItemDto = {
  id: number;
  residentGuid: string;
  episodeId?: string | null;
  residentCaseId?: string | null;
  centreEpisodeCode?: string | null;
  entryYear?: number | null;
  entryWeek?: number | null;
  entrySequence?: number | null;
  caseStatus?: string | null;
  psn: string;
  firstName: string;
  surname: string;
  nationality: string;
  age: number;
  weekNumber: number;
  roomNumber: string;
  unitId: Resident["unit"];
  photoUrl?: string | null;
  admissionDate?: string | null;
  expectedCompletion?: string | null;
  primaryAddiction?: string | null;
  isDrug?: boolean;
  isGambeler?: boolean;
  isPreviousResident?: boolean;
  dietaryNeedsCode?: number;
  isSnorer?: boolean;
  hasCriminalHistory?: boolean;
  isOnProbation?: boolean;
  argumentativeScale?: number;
  learningDifficultyScale?: number;
  literacyScale?: number;
};

export const getResidentSource = (): "api" | "mock" => residentSource;

const mapApiResident = (dto: ResidentListItemDto): Resident => {
  const now = new Date();
  const admissionDate = new Date(now);
  admissionDate.setDate(now.getDate() - ((dto.weekNumber || 1) * 7));
  const expectedCompletion = new Date(admissionDate);
  expectedCompletion.setDate(admissionDate.getDate() + 84);

  return {
    id: dto.id,
    residentGuid: dto.residentGuid ?? null,
    episodeId: dto.episodeId ?? null,
    residentCaseId: dto.residentCaseId ?? null,
    centreEpisodeCode: dto.centreEpisodeCode?.trim() ? dto.centreEpisodeCode : null,
    entryYear: dto.entryYear ?? null,
    entryWeek: dto.entryWeek ?? null,
    entrySequence: dto.entrySequence ?? null,
    caseStatus: dto.caseStatus?.trim() ? dto.caseStatus : null,
    firstName: dto.firstName,
    surname: dto.surname,
    nationality: dto.nationality,
    age: dto.age,
    weekNumber: dto.weekNumber,
    roomNumber: dto.roomNumber,
    unit: dto.unitId,
    photo: dto.photoUrl?.trim() ? dto.photoUrl.trim() : null,
    fallbackPhoto: `https://i.pravatar.cc/150?img=${((dto.id - 1) % 70) + 1}`,
    psn: dto.psn,
    admissionDate: dto.admissionDate?.trim() ? dto.admissionDate : admissionDate.toISOString().slice(0, 10),
    expectedCompletion: dto.expectedCompletion?.trim() ? dto.expectedCompletion : expectedCompletion.toISOString().slice(0, 10),
    primaryAddiction: dto.primaryAddiction?.trim() ? dto.primaryAddiction : "Alcohol",
    isDrug: dto.isDrug ?? dto.unitId === "drugs",
    isGambeler: dto.isGambeler ?? false,
    isPreviousResident: dto.isPreviousResident ?? false,
    dietaryNeedsCode: dto.dietaryNeedsCode ?? 0,
    isSnorer: dto.isSnorer ?? false,
    hasCriminalHistory: dto.hasCriminalHistory ?? false,
    isOnProbation: dto.isOnProbation ?? false,
    argumentativeScale: dto.argumentativeScale ?? 0,
    learningDifficultyScale: dto.learningDifficultyScale ?? 0,
    literacyScale: dto.literacyScale ?? 0,
  };
};

const fetchResidentsFromApi = async (unit: Resident["unit"]): Promise<Resident[]> => {
  const unitGuid = UNIT_GUIDS[unit];
  const response = await fetch(`${API_BASE_URL}/api/units/${encodeURIComponent(unitGuid)}/residents`, {
    method: "GET",
    headers: {
      Accept: "application/json",
    },
    cache: "no-store",
  });

  if (!response.ok) {
    throw new Error(`Residents API failed (${response.status})`);
  }

  const data = (await response.json()) as ResidentListItemDto[];
  return data.map(mapApiResident);
};

export const residentService = {
  async getResidents(unit: Resident["unit"]): Promise<Resident[]> {
    try {
      const residents = await fetchResidentsFromApi(unit);
      if (residents.length === 0) {
        await delay(50);
        residentSource = "mock";
        return mockResidents.filter((resident) => resident.unit === unit);
      }
      residentSource = "api";
      return residents;
    } catch {
      await delay(50);
      residentSource = "mock";
      return mockResidents.filter((resident) => resident.unit === unit);
    }
  },

  async getRollCallResidents(unit: Resident["unit"]): Promise<Resident[]> {
    return this.getResidents(unit);
  },

  async saveAttendance(records: AttendanceRecord[]): Promise<void> {
    await delay(50);
    store.attendance = [...store.attendance, ...records];
  },

  async recordDischarge(
    residentGuid: string,
    exitType: DischargeExitType,
    eventDate: string,
    reason: string | null,
    accessToken: string,
  ): Promise<RecordDischargeResult> {
    const clientEventId = crypto.randomUUID();
    const response = await fetch(`${API_BASE_URL}/api/residents/${encodeURIComponent(residentGuid)}/discharge`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Accept: "application/json",
        Authorization: `Bearer ${accessToken}`,
      },
      body: JSON.stringify({
        clientEventId,
        exitType: EXIT_TYPE_CODES[exitType],
        eventDate,
        reason: reason ?? null,
      }),
    });

    if (!response.ok) {
      const text = await response.text().catch(() => "");
      throw new Error(`Discharge failed (${response.status})${text ? `: ${text}` : ""}`);
    }

    return response.json() as Promise<RecordDischargeResult>;
  },
};
